### DISCUSSION TITLE
consider allowing C++ APIs in public libtransmission headers

### DISCUSSION THREAD

**Discussion by ckerr** [transmission/transmission#4568] [Ideas]
This is alongside the previous libtransmission C++izations, e.g. #1787

The current policy is to only allow C++ API in private libtransmission headers so that downstream code doesn't need to understand C++. This has an unfortunate side-effect that there exist `std::` ways to do things that cannot be used, e.g. `std::error_code` instead of `tr_error`, and also utilities that would have to live side-by-side for C and C++ world (e.g. most of libtransmission/utils.h).

One option is to migrate the rest of the project -- the GTK client, transmission-daemon, and the utils -- to C++ as well, then remove APIs in libtransmission (error.h, utils.h, file.h, ptrarray.h, list.h, etc.) when they overlap with C++ std:: APIs.

This will cause fewer ripple effects than it sounds. Unlike libtorrent, libtransmission never had a standalone release, so most projects downstream use it via RPC and will be unaffected. The main exception to this that I'm aware of is Transmission's Vala bindings which are used by [Fragments](https://gitlab.gnome.org/World/Fragments) and [Torrential](https://github.com/davidmhewitt/torrential) via Fragment's Vala bindings for Transmission.

It's been 7+ years since I wrote anything in Vala and don't know how well it plays (or doesn't play) with C++, so I wanted to file this issue and @'ing Fragments maintainer @haecker-felix and Torrential maintainer @davidmhewitt for feedback.

**Comment by davidmhewitt** [comment]
Vala bindings only work with C libraries, it can't bind directly to a C++ library.

While I'd be appreciative if there were still C headers for libtransmission, I also understand the importance of code modernization, so if this was to go ahead, I'd find some form of alternative. Whether that would be writing a minimal wrapper for the methods I needed to be able to bind them in Vala, or looking at alternative languages or libraries.

Thanks for keeping your downstreams in mind!

**Comment by ckerr** [comment]
Now that I think about it, Vala apps probably won't care about around-the-edges changes like utils.h, ptrarray.h, list.h, file.h, and error.h because Vala apps are probably using glib equivalents, e.g. GList or GSList instead of tr_list...

**Comment by davidmhewitt** [comment]
Yes, definitely not using any Transmission list structures.

This sounds fine on the face of it, but feel free to ping me for a review of any of the specific removals and I can check if it's being used and whether there's a viable alternative.

**Comment by ckerr** [comment]
A  followup to the above discussion -- Fragments 2.0 beta 1 is out, and it appears to have dropped the C bindings in favor of using the JSON RPC bindings :tada: https://mastodon.social/@haeckerfelix/107401693861903548

**Comment by ckerr** [comment]
Transmission 4.0.0-beta.1 will be coming out Real Soon Now.

`libtransmission/transmission.h` still has a very C-like API, but I do notice that a couple of C++isms have bled into it.

There aren't any major changes to this API in terms of C++ -- e.g. there aren't any std::vectors or strings, no templates, no C++ objects -- but I see there are some minor pieces that have bled in over time:

- using `using` instead of `typedef`
- using `std::array<>` instead of C-style arrays
- using `constexpr Foo = bar` instead of `#define FOO bar`

So I'm trying to decide how much effort should be spent to Cify `libtransmission/transmission.h` before `4.0.0` stable. If anyone is still using this for a downstream Vala / C project, please speak up and, better yet, PTAL and see if the current state is a blocker.

**Comment by ckerr** [comment]
There's not currently a deliverable for this issue, but I'm adding it to the `4.0.0` milestone anyway for tracking & visibility.

**Comment by Coeur** [comment]
Can we change the Milestone to 5.0? Doesn't seem strictly needed for 4.0.0, and seems like a lot of changes.

**Comment by GaryElshaw** [comment]
And this is the point where things get sticky. Is there a roadmap between 4.X -> 5.0? What should be considered a feature, a gui, or performance improvement? On the outside, it feels like evrything is iterative (and that's okay), but my gut tells me that iterative is the status quo, regardless of numbering per se.

What's your thinking, Charles @ckerr and Mike @mikedld?

As for me, i'm hoping the project, in this particular phase, will examine Transmission's performance. Just my experience, but peer connectivity; up and down rates, appear to be better in 3 than 4. That experience might be unique, but i'd love to see a wider benchmarking of how 4 compared with 3 is tracking, if possible.

For example, i'm still using utorrent for downloading, occasionally, because it is picking up far more peers and seeds than Transmission is. It also appears to have greater download speeds than i'm seeing in Transmission.

**Comment by ckerr** [comment]
@GaryElshaw I have been thinking a lot about this -- for example, I think we'll be able to find more peers if we support bep 55 -- but let's get 4.0.0 out the door first.

**Comment by GaryElshaw** [comment]
Thanks Charles.

**Comment by ckerr** [comment]
I _think_ this issue was created before `transmission/transmission` opened  up discussions, but looking back at this issue in preparation for 4.0.0 I think this is probably better treated as a discussion than an issue because it's more of a "keep aware of this" type topic rather than something that's actionable and can be fixed.

So I'm going to remove the `4.0.0` milestone and convert this to a discussion topic

**Comment by Coeur** [comment]
C++ isn't compatible with Swift, so I made [objc_transmission.h](https://github.com/transmission/transmission/blob/6e9455ea31761a62d67d8ca0d284920d0fc96c83/macosx/objc_transmission.h) as a non-C++ header specifically to bridge the APIs for Swift needs.

It belongs to this PR: #4540
